pyomo/core/base/matrix_constraint.py

```diff
@@ -0,0 +1,360 @@
+        expression."""
+        comp = self.parent_component()
+        index = self._index
+        return comp._lower[index]
+
+    @property
+    def upper(self):
+        """Access the upper bound of a constraint
+        expression."""
+        comp = self.parent_component()
+        index = self._index
+        return comp._upper[index]
+
+    @property
+    def equality(self):
+        """A boolean indicating whether this is an equality
+        constraint."""
+        comp = self.parent_component()
+        index = self._index
+        if (comp._lower[index] is None) or \
+           (comp._upper[index] is None):
+            return False
+        return comp._lower[index] == comp._upper[index]
+
+    @property
+    def strict_lower(self):
+        """A boolean indicating whether this constraint has
+        a strict lower bound."""
+        return False
+
+    @property
+    def strict_upper(self):
+        """A boolean indicating whether this constraint has
+        a strict upper bound."""
+        return False
+
+    def set_value(self, expr):
+        """Set the expression on this constraint."""
+        raise NotImplementedError(
+            "MatrixConstraint row elements can not be updated"
+        )
+
+@ModelComponentFactory.register(
+                   "A set of constraint expressions in Ax=b form.")
+class MatrixConstraint(collections_Mapping,
+                       IndexedConstraint):
+    """
+    Parameters
+    ----------
+    A_data : list
+        The values of the CSR format sparse matrix
+    A_indices : list
+        The column indices of the CSR format sparse matrix
+    A_indptr : list
+        The row start-stop pointers of the CSR format sparse matrix
+    lb : list
+        The list of constraint lower bounds
+    ub : list
+        The list of constraint upper bounds
+    x : list
+        The list of pyomo variables mapped to their appropriate column
+    """
+
+    def __init__(self, A_data, A_indices, A_indptr, lb, ub, x):
+
+        m = len(lb)
+        n = len(x)
+        nnz = len(A_data)
+        assert len(A_indices) == nnz
+        assert len(A_indptr) == m + 1
+        assert len(ub) == m
+        IndexedConstraint.__init__(self, Any)
+
+        self._A_data = A_data
+        self._A_indices = A_indices
+        self._A_indptr = A_indptr
+        self._lower = lb
+        self._upper = ub
+        self._x = tuple(x)
+
+    def construct(self, data=None):
+        """Construct the expression(s) for this constraint."""
+        if __debug__ and logger.isEnabledFor(logging.DEBUG):
+            logger.debug("Constructing constraint %s"
+                         % (self.name))
+        if self._constructed:
+            return
+        self._constructed = True
+
+        ref = weakref.ref(self)
+        with pyutilib.misc.PauseGC():
+            self._data = tuple(_MatrixConstraintData(i, ref)
+                               for i in xrange(len(self._lower)))
```

Comment:
Why prepopulate _data here?  You could instead overwrite `__getitem__` (since you are using collections_Mapping's dict API), or `_getitem_when_not_present()` (if you were using IndexedConstraint) in order to just create the data object on the fly when it was requested.  That way "intelligent" writers that looked at the Constraint Type could avoid the creation of the data objects altogether.

Reply:
The result of doing a big set of allocations like this with the GC disabled cuts the time from ~14s to ~4s for nrows=10,000,000. Given this significant speedup, and the fact that we don't yet have special handling of this component anywhere (so all _MatrixConstraintData objects _will_ be created), I see no need to make the implementation more complex.